In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import polars as pl

import warnings
warnings.filterwarnings('ignore')

In [2]:
data = pl.read_parquet(r"E:\Learning\TEAI Cup\Data\# Parquets\d2_60_0.parquet")

In [3]:
df = data.clone()

In [4]:
df.head()

shape: (5, 11)
┌─────────────┬─────────────┬────────────┬────────────┬───┬────────────┬──────┬───────┬────────────┐
│ device_name ┆ machine_def ┆ variable_n ┆ value      ┆ … ┆ machine_de ┆ year ┆ month ┆ date       │
│ ---         ┆ inition     ┆ ame        ┆ ---        ┆   ┆ f          ┆ ---  ┆ ---   ┆ ---        │
│ str         ┆ ---         ┆ ---        ┆ str        ┆   ┆ ---        ┆ i64  ┆ i64   ┆ str        │
│             ┆ str         ┆ str        ┆            ┆   ┆ str        ┆      ┆       ┆            │
╞═════════════╪═════════════╪════════════╪════════════╪═══╪════════════╪══════╪═══════╪════════════╡
│ de512stmp22 ┆ spark-machi ┆ teil1GapX  ┆ 0.509,0.51 ┆ … ┆ spark-mach ┆ 2025 ┆ 10    ┆ 2025-10-03 │
│ 3           ┆ ne-otto-vis ┆            ┆ 1,0.509,0. ┆   ┆ ine-otto-v ┆      ┆       ┆            │
│             ┆ ion-dyna…   ┆            ┆ 518,0.521, ┆   ┆ ision-dyna ┆      ┆       ┆            │
│             ┆             ┆            ┆ …          ┆   ┆ …          ┆      ┆       ┆            │
│ de512stmp22 ┆ spark-machi ┆ teil1GapY  ┆ 0.497,0.49 ┆ … ┆ spark-mach ┆ 2025 ┆ 10    ┆ 2025-10-03 │
│ 3           ┆ ne-otto-vis ┆            ┆ 7,0.494,0. ┆   ┆ ine-otto-v ┆      ┆       ┆            │
│             ┆ ion-dyna…   ┆            ┆ 501,0.499, ┆   ┆ ision-dyna ┆      ┆       ┆            │
│             ┆             ┆            ┆ …          ┆   ┆ …          ┆      ┆       ┆            │
│ de512stmp22 ┆ spark-machi ┆ iT1Rastfed ┆ 48.987,49. ┆ … ┆ spark-mach ┆ 2025 ┆ 10    ┆ 2025-10-03 │
│ 3           ┆ ne-otto-vis ┆ erDeformie ┆ 242,51.589 ┆   ┆ ine-otto-v ┆      ┆       ┆            │
│             ┆ ion-dyna…   ┆ rt         ┆ ,55.409,48 ┆   ┆ ision-dyna ┆      ┆       ┆            │
│             ┆             ┆            ┆ …          ┆   ┆ …          ┆      ┆       ┆            │
│ de512stmp22 ┆ spark-machi ┆ iT1Rastfed ┆ 0.014,0.01 ┆ … ┆ spark-mach ┆ 2025 ┆ 10    ┆ 2025-10-03 │
│ 3           ┆ ne-otto-vis ┆ erdifferen ┆ 7,0.013,0. ┆   ┆ ine-otto-v ┆      ┆       ┆            │
│             ┆ ion-dyna…   ┆ z          ┆ 017,0.011, ┆   ┆ ision-dyna ┆      ┆       ┆            │
│             ┆             ┆            ┆ …          ┆   ┆ …          ┆      ┆       ┆            │
│ de512stmp22 ┆ spark-machi ┆ onlineGaug ┆ 3.979,3.97 ┆ … ┆ spark-mach ┆ 2025 ┆ 10    ┆ 2025-10-03 │
│ 3           ┆ ne-otto-vis ┆ e1         ┆ 8,3.979,3. ┆   ┆ ine-otto-v ┆      ┆       ┆            │
│             ┆ ion-dyna…   ┆            ┆ 979,3.979, ┆   ┆ ision-dyna ┆      ┆       ┆            │
│             ┆             ┆            ┆ …          ┆   ┆ …          ┆      ┆       ┆            │
└─────────────┴─────────────┴────────────┴────────────┴───┴────────────┴──────┴───────┴────────────┘

In [5]:
lt = df.columns
print(lt)

['device_name', 'machine_definition', 'variable_name', 'value', 'timestamp', 'variable_attribute', 'device', 'machine_def', 'year', 'month', 'date']


## Removing unnecessory data features

In [6]:
unnecessory_features = ["device", "machine_def", "year", "month", "date"]
for feature in unnecessory_features:
    if feature in df.columns:
        df = df.drop(pl.col(feature))
df.head()

shape: (5, 6)
┌──────────────┬────────────────┬────────────────┬────────────────┬────────────────┬───────────────┐
│ device_name  ┆ machine_defini ┆ variable_name  ┆ value          ┆ timestamp      ┆ variable_attr │
│ ---          ┆ tion           ┆ ---            ┆ ---            ┆ ---            ┆ ibute         │
│ str          ┆ ---            ┆ str            ┆ str            ┆ str            ┆ ---           │
│              ┆ str            ┆                ┆                ┆                ┆ str           │
╞══════════════╪════════════════╪════════════════╪════════════════╪════════════════╪═══════════════╡
│ de512stmp223 ┆ spark-machine- ┆ teil1GapX      ┆ 0.509,0.511,0. ┆ 2025-10-03 00: ┆ lowerLimit:   │
│              ┆ otto-vision-dy ┆                ┆ 509,0.518,0.52 ┆ 11:02.091000   ┆ -0.06499,     │
│              ┆ na…            ┆                ┆ 1,…            ┆                ┆ upperLim…     │
│ de512stmp223 ┆ spark-machine- ┆ teil1GapY      ┆ 0.497,0.497,0. ┆ 2025-10-03 00: ┆ lowerLimit:   │
│              ┆ otto-vision-dy ┆                ┆ 494,0.501,0.49 ┆ 11:02.091000   ┆ -0.06499,     │
│              ┆ na…            ┆                ┆ 9,…            ┆                ┆ upperLim…     │
│ de512stmp223 ┆ spark-machine- ┆ iT1RastfederDe ┆ 48.987,49.242, ┆ 2025-10-03 00: ┆ lowerLimit:   │
│              ┆ otto-vision-dy ┆ formiert       ┆ 51.589,55.409, ┆ 11:02.091000   ┆ -20,          │
│              ┆ na…            ┆                ┆ 48…            ┆                ┆ upperLimit:   │
│              ┆                ┆                ┆                ┆                ┆ 2…            │
│ de512stmp223 ┆ spark-machine- ┆ iT1Rastfederdi ┆ 0.014,0.017,0. ┆ 2025-10-03 00: ┆ lowerLimit:   │
│              ┆ otto-vision-dy ┆ fferenz        ┆ 013,0.017,0.01 ┆ 11:02.091000   ┆ -0.08,        │
│              ┆ na…            ┆                ┆ 1,…            ┆                ┆ upperLimit:…  │
│ de512stmp223 ┆ spark-machine- ┆ onlineGauge1   ┆ 3.979,3.978,3. ┆ 2025-10-03 00: ┆ lowerLimit:   │
│              ┆ otto-vision-dy ┆                ┆ 979,3.979,3.97 ┆ 11:02.091000   ┆ -0.1,         │
│              ┆ na…            ┆                ┆ 9,…            ┆                ┆ upperLimit: … │
└──────────────┴────────────────┴────────────────┴────────────────┴────────────────┴───────────────┘

In [7]:
# Now got the data as we have previously

In [8]:
df.dtypes

[String, String, String, String, String, String]

# Converting to timestamp

In [9]:
if df["timestamp"].dtype == pl.String:
    df = df.with_columns(
        pl.col("timestamp").str.to_datetime()
    )
df.head()

shape: (5, 6)
┌──────────────┬────────────────┬────────────────┬────────────────┬────────────────┬───────────────┐
│ device_name  ┆ machine_defini ┆ variable_name  ┆ value          ┆ timestamp      ┆ variable_attr │
│ ---          ┆ tion           ┆ ---            ┆ ---            ┆ ---            ┆ ibute         │
│ str          ┆ ---            ┆ str            ┆ str            ┆ datetime[μs]   ┆ ---           │
│              ┆ str            ┆                ┆                ┆                ┆ str           │
╞══════════════╪════════════════╪════════════════╪════════════════╪════════════════╪═══════════════╡
│ de512stmp223 ┆ spark-machine- ┆ teil1GapX      ┆ 0.509,0.511,0. ┆ 2025-10-03     ┆ lowerLimit:   │
│              ┆ otto-vision-dy ┆                ┆ 509,0.518,0.52 ┆ 00:11:02.091   ┆ -0.06499,     │
│              ┆ na…            ┆                ┆ 1,…            ┆                ┆ upperLim…     │
│ de512stmp223 ┆ spark-machine- ┆ teil1GapY      ┆ 0.497,0.497,0. ┆ 2025-10-03     ┆ lowerLimit:   │
│              ┆ otto-vision-dy ┆                ┆ 494,0.501,0.49 ┆ 00:11:02.091   ┆ -0.06499,     │
│              ┆ na…            ┆                ┆ 9,…            ┆                ┆ upperLim…     │
│ de512stmp223 ┆ spark-machine- ┆ iT1RastfederDe ┆ 48.987,49.242, ┆ 2025-10-03     ┆ lowerLimit:   │
│              ┆ otto-vision-dy ┆ formiert       ┆ 51.589,55.409, ┆ 00:11:02.091   ┆ -20,          │
│              ┆ na…            ┆                ┆ 48…            ┆                ┆ upperLimit:   │
│              ┆                ┆                ┆                ┆                ┆ 2…            │
│ de512stmp223 ┆ spark-machine- ┆ iT1Rastfederdi ┆ 0.014,0.017,0. ┆ 2025-10-03     ┆ lowerLimit:   │
│              ┆ otto-vision-dy ┆ fferenz        ┆ 013,0.017,0.01 ┆ 00:11:02.091   ┆ -0.08,        │
│              ┆ na…            ┆                ┆ 1,…            ┆                ┆ upperLimit:…  │
│ de512stmp223 ┆ spark-machine- ┆ onlineGauge1   ┆ 3.979,3.978,3. ┆ 2025-10-03     ┆ lowerLimit:   │
│              ┆ otto-vision-dy ┆                ┆ 979,3.979,3.97 ┆ 00:11:02.091   ┆ -0.1,         │
│              ┆ na…            ┆                ┆ 9,…            ┆                ┆ upperLimit: … │
└──────────────┴────────────────┴────────────────┴────────────────┴────────────────┴───────────────┘

## Seperating variable_attribute Feature:
* lowerLimit: -0.06499, upperLimit: 0.00499, nominalValue: 0.53, engineeringUnits: mm, cavity: 1

In [10]:
keys = ["lowerLimit", "upperLimit", "nominalValue", "engineeringUnits", "cavity"]

df = df.with_columns(
    pl.col("variable_attribute")
    .str.split(",")
    .list.eval(
        pl.element()
        .str.split(":")
        .list.eval(
            pl.element().str.strip_chars()
        )
    )
    .list.eval(
        pl.element()
        .list.last()
    )
    .list.to_struct(fields=keys)

).unnest("variable_attribute")
df

shape: (22_702_399, 10)
┌────────────┬────────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬────────┐
│ device_nam ┆ machine_de ┆ variable_ ┆ value     ┆ … ┆ upperLimi ┆ nominalVa ┆ engineeri ┆ cavity │
│ e          ┆ finition   ┆ name      ┆ ---       ┆   ┆ t         ┆ lue       ┆ ngUnits   ┆ ---    │
│ ---        ┆ ---        ┆ ---       ┆ str       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ str    │
│ str        ┆ str        ┆ str       ┆           ┆   ┆ str       ┆ str       ┆ str       ┆        │
╞════════════╪════════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪════════╡
│ de512stmp2 ┆ spark-mach ┆ teil1GapX ┆ 0.509,0.5 ┆ … ┆ 0.00499   ┆ 0.53      ┆ mm        ┆ 1      │
│ 23         ┆ ine-otto-v ┆           ┆ 11,0.509, ┆   ┆           ┆           ┆           ┆        │
│            ┆ ision-dyna ┆           ┆ 0.518,0.5 ┆   ┆           ┆           ┆           ┆        │
│            ┆ …          ┆           ┆ 21,…      ┆   ┆           ┆           ┆           ┆        │
│ de512stmp2 ┆ spark-mach ┆ teil1GapY ┆ 0.497,0.4 ┆ … ┆ 0.00499   ┆ 0.53      ┆ mm        ┆ 1      │
│ 23         ┆ ine-otto-v ┆           ┆ 97,0.494, ┆   ┆           ┆           ┆           ┆        │
│            ┆ ision-dyna ┆           ┆ 0.501,0.4 ┆   ┆           ┆           ┆           ┆        │
│            ┆ …          ┆           ┆ 99,…      ┆   ┆           ┆           ┆           ┆        │
│ de512stmp2 ┆ spark-mach ┆ iT1Rastfe ┆ 48.987,49 ┆ … ┆ 22        ┆ 50        ┆ mm        ┆ 1      │
│ 23         ┆ ine-otto-v ┆ derDeform ┆ .242,51.5 ┆   ┆           ┆           ┆           ┆        │
│            ┆ ision-dyna ┆ iert      ┆ 89,55.409 ┆   ┆           ┆           ┆           ┆        │
│            ┆ …          ┆           ┆ ,48…      ┆   ┆           ┆           ┆           ┆        │
│ de512stmp2 ┆ spark-mach ┆ iT1Rastfe ┆ 0.014,0.0 ┆ … ┆ 0.08      ┆ 0         ┆ mm        ┆ 1      │
│ 23         ┆ ine-otto-v ┆ derdiffer ┆ 17,0.013, ┆   ┆           ┆           ┆           ┆        │
│            ┆ ision-dyna ┆ enz       ┆ 0.017,0.0 ┆   ┆           ┆           ┆           ┆        │
│            ┆ …          ┆           ┆ 11,…      ┆   ┆           ┆           ┆           ┆        │
│ de512stmp2 ┆ spark-mach ┆ onlineGau ┆ 3.979,3.9 ┆ … ┆ 0.1       ┆ 4         ┆ mm        ┆ 1      │
│ 23         ┆ ine-otto-v ┆ ge1       ┆ 78,3.979, ┆   ┆           ┆           ┆           ┆        │
│            ┆ ision-dyna ┆           ┆ 3.979,3.9 ┆   ┆           ┆           ┆           ┆        │
│            ┆ …          ┆           ┆ 79,…      ┆   ┆           ┆           ┆           ┆        │
│ …          ┆ …          ┆ …         ┆ …         ┆ … ┆ …         ┆ …         ┆ …         ┆ …      │
│ de512stmp2 ┆ spark-mach ┆ iso       ┆ 4.225,4.2 ┆ … ┆ 0.2       ┆ 4.2       ┆ mm        ┆ 1      │
│ 76         ┆ ine-otto-v ┆           ┆ 21,4.244, ┆   ┆           ┆           ┆           ┆        │
│            ┆ ision-dyna ┆           ┆ 4.241,4.2 ┆   ┆           ┆           ┆           ┆        │
│            ┆ …          ┆           ┆ 32,…      ┆   ┆           ┆           ┆           ┆        │
│ de512stmp2 ┆ spark-mach ┆ sp1DlErke ┆ 24489.215 ┆ … ┆ 65000     ┆ 0         ┆ mm        ┆ 1      │
│ 76         ┆ ine-otto-v ┆ nnungErge ┆ ,26002.90 ┆   ┆           ┆           ┆           ┆        │
│            ┆ ision-dyna ┆ bnis      ┆ 8,20287.1 ┆   ┆           ┆           ┆           ┆        │
│            ┆ …          ┆           ┆ 36,…      ┆   ┆           ┆           ┆           ┆        │
│ de512stmp2 ┆ spark-mach ┆ sp2DlErke ┆ 34480.418 ┆ … ┆ 55000     ┆ 0         ┆ mm        ┆ 1      │
│ 76         ┆ ine-otto-v ┆ nnungErge ┆ ,33712.51 ┆   ┆           ┆           ┆           ┆        │
│            ┆ ision-dyna ┆ bnis      ┆ 8,36825.8 ┆   ┆           ┆           ┆           ┆        │
│            ┆ …          ┆           ┆ 86,…      ┆   ┆           ┆           ┆           ┆        │
│ de512stmp2 ┆ spark-mach ┆ l-tstelle

## Removing unnecessory data features

In [11]:
unnecessory_features = ["engineeringUnits", "cavity", "machine_definition"]
for feature in unnecessory_features:
    if feature in df.columns:
        df = df.drop(pl.col(feature))
df.head()

shape: (5, 7)
┌──────────────┬──────────────┬──────────────┬─────────────┬────────────┬────────────┬─────────────┐
│ device_name  ┆ variable_nam ┆ value        ┆ timestamp   ┆ lowerLimit ┆ upperLimit ┆ nominalValu │
│ ---          ┆ e            ┆ ---          ┆ ---         ┆ ---        ┆ ---        ┆ e           │
│ str          ┆ ---          ┆ str          ┆ datetime[μs ┆ str        ┆ str        ┆ ---         │
│              ┆ str          ┆              ┆ ]           ┆            ┆            ┆ str         │
╞══════════════╪══════════════╪══════════════╪═════════════╪════════════╪════════════╪═════════════╡
│ de512stmp223 ┆ teil1GapX    ┆ 0.509,0.511, ┆ 2025-10-03  ┆ -0.06499   ┆ 0.00499    ┆ 0.53        │
│              ┆              ┆ 0.509,0.518, ┆ 00:11:02.09 ┆            ┆            ┆             │
│              ┆              ┆ 0.521,…      ┆ 1           ┆            ┆            ┆             │
│ de512stmp223 ┆ teil1GapY    ┆ 0.497,0.497, ┆ 2025-10-03  ┆ -0.06499   ┆ 0.00499    ┆ 0.53        │
│              ┆              ┆ 0.494,0.501, ┆ 00:11:02.09 ┆            ┆            ┆             │
│              ┆              ┆ 0.499,…      ┆ 1           ┆            ┆            ┆             │
│ de512stmp223 ┆ iT1Rastfeder ┆ 48.987,49.24 ┆ 2025-10-03  ┆ -20        ┆ 22         ┆ 50          │
│              ┆ Deformiert   ┆ 2,51.589,55. ┆ 00:11:02.09 ┆            ┆            ┆             │
│              ┆              ┆ 409,48…      ┆ 1           ┆            ┆            ┆             │
│ de512stmp223 ┆ iT1Rastfeder ┆ 0.014,0.017, ┆ 2025-10-03  ┆ -0.08      ┆ 0.08       ┆ 0           │
│              ┆ differenz    ┆ 0.013,0.017, ┆ 00:11:02.09 ┆            ┆            ┆             │
│              ┆              ┆ 0.011,…      ┆ 1           ┆            ┆            ┆             │
│ de512stmp223 ┆ onlineGauge1 ┆ 3.979,3.978, ┆ 2025-10-03  ┆ -0.1       ┆ 0.1        ┆ 4           │
│              ┆              ┆ 3.979,3.979, ┆ 00:11:02.09 ┆            ┆            ┆             │
│              ┆              ┆ 3.979,…      ┆ 1           ┆            ┆            ┆             │
└──────────────┴──────────────┴──────────────┴─────────────┴────────────┴────────────┴─────────────┘

## Making new features float

In [12]:
df = df.with_columns(
    pl.col(['lowerLimit', 'upperLimit', 'nominalValue']).cast(pl.Float64)
)
df

shape: (22_702_399, 7)
┌──────────────┬──────────────┬──────────────┬─────────────┬────────────┬────────────┬─────────────┐
│ device_name  ┆ variable_nam ┆ value        ┆ timestamp   ┆ lowerLimit ┆ upperLimit ┆ nominalValu │
│ ---          ┆ e            ┆ ---          ┆ ---         ┆ ---        ┆ ---        ┆ e           │
│ str          ┆ ---          ┆ str          ┆ datetime[μs ┆ f64        ┆ f64        ┆ ---         │
│              ┆ str          ┆              ┆ ]           ┆            ┆            ┆ f64         │
╞══════════════╪══════════════╪══════════════╪═════════════╪════════════╪════════════╪═════════════╡
│ de512stmp223 ┆ teil1GapX    ┆ 0.509,0.511, ┆ 2025-10-03  ┆ -0.06499   ┆ 0.00499    ┆ 0.53        │
│              ┆              ┆ 0.509,0.518, ┆ 00:11:02.09 ┆            ┆            ┆             │
│              ┆              ┆ 0.521,…      ┆ 1           ┆            ┆            ┆             │
│ de512stmp223 ┆ teil1GapY    ┆ 0.497,0.497, ┆ 2025-10-03  ┆ -0.06499   ┆ 0.00499    ┆ 0.53        │
│              ┆              ┆ 0.494,0.501, ┆ 00:11:02.09 ┆            ┆            ┆             │
│              ┆              ┆ 0.499,…      ┆ 1           ┆            ┆            ┆             │
│ de512stmp223 ┆ iT1Rastfeder ┆ 48.987,49.24 ┆ 2025-10-03  ┆ -20.0      ┆ 22.0       ┆ 50.0        │
│              ┆ Deformiert   ┆ 2,51.589,55. ┆ 00:11:02.09 ┆            ┆            ┆             │
│              ┆              ┆ 409,48…      ┆ 1           ┆            ┆            ┆             │
│ de512stmp223 ┆ iT1Rastfeder ┆ 0.014,0.017, ┆ 2025-10-03  ┆ -0.08      ┆ 0.08       ┆ 0.0         │
│              ┆ differenz    ┆ 0.013,0.017, ┆ 00:11:02.09 ┆            ┆            ┆             │
│              ┆              ┆ 0.011,…      ┆ 1           ┆            ┆            ┆             │
│ de512stmp223 ┆ onlineGauge1 ┆ 3.979,3.978, ┆ 2025-10-03  ┆ -0.1       ┆ 0.1        ┆ 4.0         │
│              ┆              ┆ 3.979,3.979, ┆ 00:11:02.09 ┆            ┆            ┆             │
│              ┆              ┆ 3.979,…      ┆ 1           ┆            ┆            ┆             │
│ …            ┆ …            ┆ …            ┆ …           ┆ …          ┆ …          ┆ …           │
│ de512stmp276 ┆ iso          ┆ 4.225,4.221, ┆ 2025-12-01  ┆ -0.2       ┆ 0.2        ┆ 4.2         │
│              ┆              ┆ 4.244,4.241, ┆ 22:22:57.45 ┆            ┆            ┆             │
│              ┆              ┆ 4.232,…      ┆ 7           ┆            ┆            ┆             │
│ de512stmp276 ┆ sp1DlErkennu ┆ 24489.215,26 ┆ 2025-12-01  ┆ -0.1       ┆ 65000.0    ┆ 0.0         │
│              ┆ ngErgebnis   ┆ 002.908,2028 ┆ 22:22:57.45 ┆            ┆            ┆             │
│              ┆              ┆ 7.136,…      ┆ 7           ┆            ┆            ┆             │
│ de512stmp276 ┆ sp2DlErkennu ┆ 34480.418,33 ┆ 2025-12-01  ┆ -0.1       ┆ 55000.0    ┆ 0.0         │
│              ┆ ngErgebnis   ┆ 712.518,3682 ┆ 22:22:57.45 ┆            ┆            ┆             │
│              ┆              ┆ 5.886,…      ┆ 7           ┆            ┆            ┆             │
│ de512stmp276 ┆ l-tstelleBod ┆ 2,2,2,2,2,2, ┆ 2025-12-01  ┆ -1.0       ┆ 1.0        ┆ 2.0         │
│              ┆ y            ┆ 2,2,2,2,2,2, ┆ 22:22:57.45 ┆            ┆            ┆             │
│              ┆              ┆ 2,2,2,…      ┆ 7           ┆            ┆            ┆             │
│ de512stmp276 ┆ l-tstelleFed ┆ 2,2,2,2,2,2, ┆ 2025-12-01  ┆ -1.0       ┆ 1.0        ┆ 2.0         │
│              ┆ er           ┆ 2,2,2,2,2,2, ┆ 22:22:57.45 ┆            ┆            ┆             │
│              ┆              ┆ 2,2,2,…      ┆ 7           ┆            ┆            ┆             │
└──────────────┴──────────────┴──────────────┴─────────────┴────────────┴────────────┴─────────────┘

## Converting the device name to name matching other datasets


In [13]:
df = df.with_columns((
    "S-" + 
    pl.col("device_name")
    .str.slice(-3))
    .alias("device_name")
)
df

shape: (22_702_399, 7)
┌─────────────┬──────────────┬──────────────┬──────────────┬────────────┬────────────┬─────────────┐
│ device_name ┆ variable_nam ┆ value        ┆ timestamp    ┆ lowerLimit ┆ upperLimit ┆ nominalValu │
│ ---         ┆ e            ┆ ---          ┆ ---          ┆ ---        ┆ ---        ┆ e           │
│ str         ┆ ---          ┆ str          ┆ datetime[μs] ┆ f64        ┆ f64        ┆ ---         │
│             ┆ str          ┆              ┆              ┆            ┆            ┆ f64         │
╞═════════════╪══════════════╪══════════════╪══════════════╪════════════╪════════════╪═════════════╡
│ S-223       ┆ teil1GapX    ┆ 0.509,0.511, ┆ 2025-10-03   ┆ -0.06499   ┆ 0.00499    ┆ 0.53        │
│             ┆              ┆ 0.509,0.518, ┆ 00:11:02.091 ┆            ┆            ┆             │
│             ┆              ┆ 0.521,…      ┆              ┆            ┆            ┆             │
│ S-223       ┆ teil1GapY    ┆ 0.497,0.497, ┆ 2025-10-03   ┆ -0.06499   ┆ 0.00499    ┆ 0.53        │
│             ┆              ┆ 0.494,0.501, ┆ 00:11:02.091 ┆            ┆            ┆             │
│             ┆              ┆ 0.499,…      ┆              ┆            ┆            ┆             │
│ S-223       ┆ iT1Rastfeder ┆ 48.987,49.24 ┆ 2025-10-03   ┆ -20.0      ┆ 22.0       ┆ 50.0        │
│             ┆ Deformiert   ┆ 2,51.589,55. ┆ 00:11:02.091 ┆            ┆            ┆             │
│             ┆              ┆ 409,48…      ┆              ┆            ┆            ┆             │
│ S-223       ┆ iT1Rastfeder ┆ 0.014,0.017, ┆ 2025-10-03   ┆ -0.08      ┆ 0.08       ┆ 0.0         │
│             ┆ differenz    ┆ 0.013,0.017, ┆ 00:11:02.091 ┆            ┆            ┆             │
│             ┆              ┆ 0.011,…      ┆              ┆            ┆            ┆             │
│ S-223       ┆ onlineGauge1 ┆ 3.979,3.978, ┆ 2025-10-03   ┆ -0.1       ┆ 0.1        ┆ 4.0         │
│             ┆              ┆ 3.979,3.979, ┆ 00:11:02.091 ┆            ┆            ┆             │
│             ┆              ┆ 3.979,…      ┆              ┆            ┆            ┆             │
│ …           ┆ …            ┆ …            ┆ …            ┆ …          ┆ …          ┆ …           │
│ S-276       ┆ iso          ┆ 4.225,4.221, ┆ 2025-12-01   ┆ -0.2       ┆ 0.2        ┆ 4.2         │
│             ┆              ┆ 4.244,4.241, ┆ 22:22:57.457 ┆            ┆            ┆             │
│             ┆              ┆ 4.232,…      ┆              ┆            ┆            ┆             │
│ S-276       ┆ sp1DlErkennu ┆ 24489.215,26 ┆ 2025-12-01   ┆ -0.1       ┆ 65000.0    ┆ 0.0         │
│             ┆ ngErgebnis   ┆ 002.908,2028 ┆ 22:22:57.457 ┆            ┆            ┆             │
│             ┆              ┆ 7.136,…      ┆              ┆            ┆            ┆             │
│ S-276       ┆ sp2DlErkennu ┆ 34480.418,33 ┆ 2025-12-01   ┆ -0.1       ┆ 55000.0    ┆ 0.0         │
│             ┆ ngErgebnis   ┆ 712.518,3682 ┆ 22:22:57.457 ┆            ┆            ┆             │
│             ┆              ┆ 5.886,…      ┆              ┆            ┆            ┆             │
│ S-276       ┆ l-tstelleBod ┆ 2,2,2,2,2,2, ┆ 2025-12-01   ┆ -1.0       ┆ 1.0        ┆ 2.0         │
│             ┆ y            ┆ 2,2,2,2,2,2, ┆ 22:22:57.457 ┆            ┆            ┆             │
│             ┆              ┆ 2,2,2,…      ┆              ┆            ┆            ┆             │
│ S-276       ┆ l-tstelleFed ┆ 2,2,2,2,2,2, ┆ 2025-12-01   ┆ -1.0       ┆ 1.0        ┆ 2.0         │
│             ┆ er           ┆ 2,2,2,2,2,2, ┆ 22:22:57.457 ┆            ┆            ┆             │
│             ┆              ┆ 2,2,2,…      ┆              ┆            ┆            ┆             │
└─────────────┴──────────────┴──────────────┴──────────────┴────────────┴────────────┴─────────────┘

### Renaming the device_name : machine_id

In [14]:
df = df.rename({'device_name': 'machine_id'})
df

shape: (22_702_399, 7)
┌────────────┬──────────────┬──────────────┬──────────────┬────────────┬────────────┬──────────────┐
│ machine_id ┆ variable_nam ┆ value        ┆ timestamp    ┆ lowerLimit ┆ upperLimit ┆ nominalValue │
│ ---        ┆ e            ┆ ---          ┆ ---          ┆ ---        ┆ ---        ┆ ---          │
│ str        ┆ ---          ┆ str          ┆ datetime[μs] ┆ f64        ┆ f64        ┆ f64          │
│            ┆ str          ┆              ┆              ┆            ┆            ┆              │
╞════════════╪══════════════╪══════════════╪══════════════╪════════════╪════════════╪══════════════╡
│ S-223      ┆ teil1GapX    ┆ 0.509,0.511, ┆ 2025-10-03   ┆ -0.06499   ┆ 0.00499    ┆ 0.53         │
│            ┆              ┆ 0.509,0.518, ┆ 00:11:02.091 ┆            ┆            ┆              │
│            ┆              ┆ 0.521,…      ┆              ┆            ┆            ┆              │
│ S-223      ┆ teil1GapY    ┆ 0.497,0.497, ┆ 2025-10-03   ┆ -0.06499   ┆ 0.00499    ┆ 0.53         │
│            ┆              ┆ 0.494,0.501, ┆ 00:11:02.091 ┆            ┆            ┆              │
│            ┆              ┆ 0.499,…      ┆              ┆            ┆            ┆              │
│ S-223      ┆ iT1Rastfeder ┆ 48.987,49.24 ┆ 2025-10-03   ┆ -20.0      ┆ 22.0       ┆ 50.0         │
│            ┆ Deformiert   ┆ 2,51.589,55. ┆ 00:11:02.091 ┆            ┆            ┆              │
│            ┆              ┆ 409,48…      ┆              ┆            ┆            ┆              │
│ S-223      ┆ iT1Rastfeder ┆ 0.014,0.017, ┆ 2025-10-03   ┆ -0.08      ┆ 0.08       ┆ 0.0          │
│            ┆ differenz    ┆ 0.013,0.017, ┆ 00:11:02.091 ┆            ┆            ┆              │
│            ┆              ┆ 0.011,…      ┆              ┆            ┆            ┆              │
│ S-223      ┆ onlineGauge1 ┆ 3.979,3.978, ┆ 2025-10-03   ┆ -0.1       ┆ 0.1        ┆ 4.0          │
│            ┆              ┆ 3.979,3.979, ┆ 00:11:02.091 ┆            ┆            ┆              │
│            ┆              ┆ 3.979,…      ┆              ┆            ┆            ┆              │
│ …          ┆ …            ┆ …            ┆ …            ┆ …          ┆ …          ┆ …            │
│ S-276      ┆ iso          ┆ 4.225,4.221, ┆ 2025-12-01   ┆ -0.2       ┆ 0.2        ┆ 4.2          │
│            ┆              ┆ 4.244,4.241, ┆ 22:22:57.457 ┆            ┆            ┆              │
│            ┆              ┆ 4.232,…      ┆              ┆            ┆            ┆              │
│ S-276      ┆ sp1DlErkennu ┆ 24489.215,26 ┆ 2025-12-01   ┆ -0.1       ┆ 65000.0    ┆ 0.0          │
│            ┆ ngErgebnis   ┆ 002.908,2028 ┆ 22:22:57.457 ┆            ┆            ┆              │
│            ┆              ┆ 7.136,…      ┆              ┆            ┆            ┆              │
│ S-276      ┆ sp2DlErkennu ┆ 34480.418,33 ┆ 2025-12-01   ┆ -0.1       ┆ 55000.0    ┆ 0.0          │
│            ┆ ngErgebnis   ┆ 712.518,3682 ┆ 22:22:57.457 ┆            ┆            ┆              │
│            ┆              ┆ 5.886,…      ┆              ┆            ┆            ┆              │
│ S-276      ┆ l-tstelleBod ┆ 2,2,2,2,2,2, ┆ 2025-12-01   ┆ -1.0       ┆ 1.0        ┆ 2.0          │
│            ┆ y            ┆ 2,2,2,2,2,2, ┆ 22:22:57.457 ┆            ┆            ┆              │
│            ┆              ┆ 2,2,2,…      ┆              ┆            ┆            ┆              │
│ S-276      ┆ l-tstelleFed ┆ 2,2,2,2,2,2, ┆ 2025-12-01   ┆ -1.0       ┆ 1.0        ┆ 2.0          │
│            ┆ er           ┆ 2,2,2,2,2,2, ┆ 22:22:57.457 ┆            ┆            ┆              │
│            ┆              ┆ 2,2,2,…      ┆              ┆            ┆            ┆              │
└────────────┴──────────────┴──────────────┴──────────────┴────────────┴────────────┴──────────────┘

In [15]:
df = df.with_columns(
    pl.col("value").str.extract_all(r"-?\d+\.?\d*").cast(pl.List(pl.Float64))
)
df

: 

# Adding some feature to dataset to detect the variations in value field

In [15]:
lf = df.lazy()
lf

In [16]:
vals = (
    pl.col('value')
    .str.split(",")
    .list.eval(
        pl.element().cast(
            pl.Float64
        )
    )
)

In [17]:
outside_range_checks = [
    (vals.list.get(i) > pl.col("upperLimit"))
    | (vals.list.get(i) < pl.col("lowerLimit"))
    for i in range(100)
]


In [18]:
lf = lf.with_columns(
    [
        pl.col("value")
        .str.extract_all(r"-?\d+\.?\d*")
        .cast(pl.List(pl.Float64))
        .list.min()
        .alias("min")
    ]
)


In [19]:
df = lf.collect(streaming=True)
df

: 